colossalai/shardformer/policies/bert.py

```diff
@@ -251,13 +251,34 @@ def set_pipeline_forward(self, model_cls: nn.Module, new_forward: Callable, poli
             else:
                 module = self.model.bert
 
-            layers_per_stage = Policy.distribute_layers(len(module.encoder.layer), stage_manager.num_stages)
-            stage_index = Policy.get_stage_index(layers_per_stage, stage_manager.stage)
-            method_replacement = {
-                "forward": partial(
-                    new_forward, stage_manager=stage_manager, stage_index=stage_index, shard_config=self.shard_config
+            # num_model_chunks > 1 if interleaved
+            num_model_chunks = stage_manager.num_model_chunks
+            layers_per_stage = Policy.distribute_layers(
+                len(module.encoder.layer), stage_manager.num_stages * num_model_chunks
+            )
+
+            if num_model_chunks > 1:
+                stage_index = Policy.get_stage_index(
+                    layers_per_stage, stage_manager.stage, stage_manager.num_stages, num_model_chunks
                 )
-            }
+            else:
+                stage_index = Policy.get_stage_index(layers_per_stage, stage_manager.stage)
+
+            if num_model_chunks == 1:
+                method_replacement = {
+                    "forward": partial(
+                        new_forward,
+                        stage_manager=stage_manager,
+                        stage_index=stage_index,
+                        shard_config=self.shard_config,
+                    )
+                }
+            # for interleaved, stage index for each forward is chosen in scheduler
+            else:
+                self.scheduler.layers = stage_index
```

Comment:
Can it be derived from stage and num model chunks?

Reply:
yes, that was actually my initial implementation. I changed it to avoid the need to modify the forward function of all other shardformer models.

if we want to sync the implementation style for interleaved and non-interleaved layer assignment for the foward function, I think the better way is to stop passing layer information as part of the partial function, since we no longer know it in advance. I can raise another PR for non-interleaved 1f1b schedule to pass layer information in the scheduler instead if that works, something like https://github.com/hpcaitech/ColossalAI/blob/ad33868e8b8d9359b1b7eca20da27583fa2d6f32/colossalai/pipeline/schedule/interleaved_pp.py#L216